macaw-x86-cli/src/Data/Macaw/X86/Symbolic/CLI.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Macaw.X86.Symbolic.CLI
+  ( withX86Hooks
+  ) where
+
+
+import Data.Proxy (Proxy(..))
+
+import Data.Parameterized.NatRepr (knownNat)
+
+import What4.Expr (ExprBuilder)
+
+import Lang.Crucible.Backend qualified as C
+import Lang.Crucible.CLI (SimulateProgramHooks, defaultSimulateProgramHooks)
+import Lang.Crucible.FunctionHandle (newHandleAllocator)
+import Lang.Crucible.LLVM.DataLayout (EndianForm(LittleEndian))
+import Lang.Crucible.LLVM.MemModel qualified as Mem
+import Lang.Crucible.Simulator.ExecutionTree (ExtensionImpl)
+import Lang.Crucible.Syntax.Concrete (ParserHooks)
+
+import Data.Macaw.CFG qualified as DMC
+import Data.Macaw.Memory qualified as DMM
+import Data.Macaw.Symbolic qualified as DMS
+import Data.Macaw.Symbolic.Memory qualified as DMS
+import Data.Macaw.Symbolic.Syntax (machineCodeParserHooks)
+import Data.Macaw.X86 (X86_64)
+import Data.Macaw.X86.Symbolic (newSymFuns, x86_64MacawEvalFn)
+import Data.Macaw.X86.Symbolic.Syntax (x86ParserHooks)
+
+ -- | Small helper for providing LLVM language-specific hooks, e.g., for use with
+-- 'Lang.Crucible.CLI.execCommand'.
+withX86Hooks ::
+  ((?parserHooks :: ParserHooks (DMS.MacawExt X86_64)) =>
+   (forall sym bak t st fs. 
+     ( C.IsSymBackend sym bak
+     , sym ~ ExprBuilder t st fs
+     ) =>
+     bak ->
+     IO (ExtensionImpl () sym (DMS.MacawExt X86_64))) ->
+    SimulateProgramHooks (DMS.MacawExt X86_64) ->
+    IO a) ->
+  IO a
+withX86Hooks k = do
+  ha <- newHandleAllocator
+  mvar <- Mem.mkMemVar "llvm_memory" ha
+  let ?ptrWidth = knownNat @64
+  let ?memOpts = Mem.defaultMemOptions
+  let ext ::
+        forall sym bak t st fs.
+        (C.IsSymBackend sym bak, sym ~ ExprBuilder t st fs) =>
+        bak ->
+        IO (ExtensionImpl () sym (DMS.MacawExt X86_64))
+      ext bak =  do
+        let sym = C.backendGetSym bak
+        let ?recordLLVMAnnotation = \_ _ _ -> pure ()
+        symFns <- newSymFuns sym
+        let elfMem = DMC.emptyMemory DMM.Addr64
+        let eFn = x86_64MacawEvalFn symFns DMS.defaultMacawArchStmtExtensionOverride
+        (_initMem, ptrTable) <-
+          DMS.newGlobalMemory
+            (Proxy @X86_64)
+            bak
+            LittleEndian
+            DMS.ConcreteMutable
+            elfMem
+        -- TOOD?
+        -- C.writeGlobal mvar initMem
```

Comment:
Yeah, this is almost certainly a good idea, but I'm not sure how to do it right now. This is all happening in `IO`, whereas `writeGlobal` is in `OverrideSim`. In `crucible-llvm-cli`, setting up the initial LLVM memory [happens in the `simulationHooks`](https://github.com/GaloisInc/crucible/blob/93bfa7f7858eb458f83eb3e0f030bf600ef72365/crucible-llvm-cli/src/Lang/Crucible/LLVM/CLI.hs#L59), but here, we get the value of the initial memory back inside a different callback (the one that returns the `ExtensionImpl`).

I think the fundamental problem here is that the `crucible-{cli,syntax}` APIs are pretty callback-oriented (based on providing "hooks"). This design kind of makes sense, because it is a big pain to set up and run the Crucible simulator, and so this design allows sharing a lot of that code. But it's not very flexible, as demonstrated by this instance.

Reply:
Perhaps it would be worth thinking about a better design and opening a `crucible` issue to discuss alternatives. Fortunately, this part of the `crucible-syntax` API is not very widely used, so I think we are at liberty to make sweeping changes to its API.

Reply:
I can't even stash the memory in an `IORef` and read it in `setupHooks`, because the type of `Mem` mentions `sym`, and all the callbacks are parametric over `sym`...

Reply:
I've opened an issue: https://github.com/GaloisInc/crucible/issues/1240

Will reference that issue in this TODO.